File: src/services/student.service.ts
export interface Student {
  id: string;
  user_id?: string;
  student_id: string;
  school_id: string;
  temporary_password?: string;
  student_name: string;
  admission_number?: string;
  grade?: string;
  section?: string;
  date_of_birth?: string;
  gender?: string;
  phone?: string;
  email?: string;
  status: 'active' | 'inactive' | 'pending';
  created_at?: string;
  updated_at?: string;
  schools?: {
    school_name: string;
    school_code: string;
  };
  assignmentStatus?: 'Assigned' | 'Unassigned';
  assigned_teacher?: string | null;
  assigned_teacher_id?: string | null;
  created_by_teacher?: string | null;
}

const API_URL = 'http://127.0.0.1:8000/api/students';

export interface StudentFilters {
  school_id?: string;
  teacher_id?: string;
  grade?: string;
  section?: string;
  status?: string;
  search?: string;
}

export interface StudentStats {
  totalStudents: number;
  activeStudents: number;
  inactiveStudents: number;
  assignedStudents: number;
  unassignedStudents: number;
  schoolDistribution: {
    schoolName: string;
    totalStudents: number;
  }[];
}

export interface StudentDetails extends Student {
  assessmentSummary: {
    completionPercentage: number;
    latestAssessmentDate: string | null;
  };
  riskOverview: {
    learningDifficultiesRisk: string;
    dyslexiaIndicators: string;
    readingFluencyProblems: string;
    attentionInconsistency: string;
    concentrationProblems: string;
    cognitiveOverload: string;
    riskLevel: string;
  };
  learningJourneySummary: {
    currentJourney: string | null;
    completionPercentage: number;
    activitiesCompleted: number;
  };
}

export const studentService = {
  createStudent: async (data: Partial<Student>): Promise<Student> => {
    const response = await fetch(API_URL, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data)
    });
    const result = await response.json();
    if (!response.ok) throw new Error(result.detail || 'Failed to create student');
    return result.student;
  },

  getStudents: async (filters?: StudentFilters): Promise<Student[]> => {
    let url = API_URL;
    if (filters) {
      const params = new URLSearchParams();
      Object.entries(filters).forEach(([key, value]) => {
        if (value) params.append(key, value);
      });
      const queryString = params.toString();
      if (queryString) url += `?${queryString}`;
    }
    const response = await fetch(url);
    if (!response.ok) throw new Error('Failed to fetch students');
    return await response.json();
  },

  getStudentById: async (id: string): Promise<StudentDetails> => {
    const response = await fetch(`${API_URL}/${id}`);
    if (!response.ok) throw new Error('Failed to fetch student details');
    return await response.json();
  },

  getStudentStats: async (): Promise<StudentStats> => {
    const response = await fetch(`${API_URL}/stats`);
    if (!response.ok) throw new Error('Failed to fetch student stats');
    return await response.json();
  },

  getStudentsBySchool: async (schoolId: string): Promise<Student[]> => {
    const response = await fetch(`${API_URL}/school/${schoolId}`);
    if (!response.ok) throw new Error('Failed to fetch students for school');
    return await response.json();
  },

  updateStudent: async (id: string, data: Partial<Student>): Promise<Student> => {
    const response = await fetch(`${API_URL}/${id}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data)
    });
    const result = await response.json();
    if (!response.ok) throw new Error(result.detail || 'Failed to update student');
    return result;
  },

  deactivateStudent: async (id: string): Promise<Student> => {
    const response = await fetch(`${API_URL}/${id}/status`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ status: 'inactive' })
    });
    const result = await response.json();
    if (!response.ok) throw new Error(result.detail || 'Failed to deactivate student');
    return result;
  }
};

File: backend/services/student_service.py
from typing import List, Dict, Any, Optional
from services.supabase_client import get_supabase_client
from datetime import datetime, timezone

class StudentService:
    @staticmethod
    def create_student(data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new student profile"""
        try:
            supabase = get_supabase_client()
            student_data = {
                "school_id": data.get("school_id"),
                "student_id": data.get("student_id"),
                "temporary_password": data.get("temporary_password"),
                "student_name": data.get("student_name"),
                "admission_number": data.get("admission_number"),
                "grade": data.get("grade"),
                "section": data.get("section"),
                "date_of_birth": data.get("date_of_birth"),
                "gender": data.get("gender"),
                "phone": data.get("phone"),
                "email": data.get("email"),
                "status": data.get("status", "active")
            }
            
            response = supabase.table('student_profiles').insert(student_data).execute()
            if not response.data:
                raise Exception("Failed to create student")
                
            return response.data[0]
        except Exception as e:
            print(f"Error creating student: {e}")
            raise Exception(f"Failed to create student: {str(e)}")

    @staticmethod
    def get_students() -> List[Dict[str, Any]]:
        """Get all students"""
        try:
            supabase = get_supabase_client()
            response = supabase.table('student_profiles').select(
                '*, schools(school_name, school_code)'
            ).order('created_at', desc=True).execute()
            return response.data
        except Exception as e:
            print(f"Error fetching students: {e}")
            return []

    @staticmethod
    def get_student_by_id(student_id: str) -> Optional[Dict[str, Any]]:
        """Get a student by ID"""
        try:
            supabase = get_supabase_client()
            response = supabase.table('student_profiles').select(
                '*, schools(school_name, school_code)'
            ).eq('id', student_id).execute()
            
            if response.data:
                return response.data[0]
            return None
        except Exception as e:
            print(f"Error fetching student: {e}")
            return None

    @staticmethod
    def get_students_by_school(school_id: str) -> List[Dict[str, Any]]:
        """Get all students in a specific school"""
        try:
            supabase = get_supabase_client()
            response = supabase.table('student_profiles').select(
                '*, schools(school_name, school_code)'
            ).eq('school_id', school_id).execute()
            return response.data
        except Exception as e:
            print(f"Error fetching students by school: {e}")
            return []

    @staticmethod
    def update_student(student_id: str, data: Dict[str, Any]) -> Dict[str, Any]:
        """Update a student"""
        try:
            supabase = get_supabase_client()
            data["updated_at"] = datetime.now(timezone.utc).isoformat()
            response = supabase.table('student_profiles').update(data).eq('id', student_id).execute()
            if not response.data:
                raise Exception("Student not found or update failed")
            return response.data[0]
        except Exception as e:
            print(f"Error updating student: {e}")
            raise Exception(f"Failed to update student: {str(e)}")

    @staticmethod
    def deactivate_student(student_id: str) -> Dict[str, Any]:
        """Soft delete a student by setting status to inactive"""
        return StudentService.update_student(student_id, {"status": "inactive"})

student_service = StudentService()
